### ISSUE TITLE
Job picked up from the queue multiple times over a longer time

### ISSUE THREAD

**Issue by RemcodM** [sidekiq/sidekiq#6379]
Ruby version: 3.3.4
Rails version: 7.1 stable branch (`7-1-stable`)
Sidekiq / Pro / Enterprise version(s): 7.3.0 / 7.3.0 / 7.3.0

We have noticed that Sidekiq executes some jobs multiple times. We started logging some of these jobs to gain a better understanding of what is happening.

This is what we know so far:
 - If the job is picked up from the queue multiple times, it always runs across different processes. Never within the same process on a different thead.
 - The job is started multiple times with exactly the same job ID. Hence we know it is the same job.
 - When the job is started, we additionally log the value of `Thread.current[:sidekiq_capsule]&.name`. This has shown us that the capsule for the thread is irrelevant. Sometimes the job is started multiple times with the same capsule name, sometimes with different capsule names. Or a combination of each.
 - Sometimes, the `start` message is logged minutes apart. Sometimes the jobs are started basically at the same time (so 2 different `start` messages are logged from different processes at the same time).
 - Sometimes a job is started 2 times, but we also found instances in which the job is started 3 times across 3 different processes.
 
Our Sidekiq configuration has Superfetch enabled. We use multiple capsules.

Our `sidekiq.yml`:
```yaml
---
# This is not the final concurreny. There is some extra concurrency
# through the capsules defined in config/initializers/sidekiq.rb.
:concurrency: 4
:queues:
  - [mb2_synchronization, 7]
  - [email_fetcher, 6]
  - [mb2, 5]
  - [mb2_background, 4]
  - [mb2_journal_entries, 4]
  - [mb2_eventually, 3]
  - [mb2_auto_link_financial_mutation_jobs, 3]
  - [mb2_subgoal_jobs, 3]
  - [open_search, 2]
  - [mb2_webhooks, 2]
  - [event_tracking, 1]
```

Relevant configured capsules in `config/initializers/sidekiq.rb`:
```ruby
  config.capsule("throughput") do |capsule|
    capsule.concurrency = 6
    capsule.queues = config[:queues].filter {|name, _| name != "mb2_journal_entries" }
  end

  config.capsule("auto_link") do |capsule|
    capsule.concurrency = 1
    capsule.queues = %w[mb2_auto_link_financial_mutation_jobs]
  end

  if ENV['APP_WORKLOAD'] == "static"
    config.capsule("throttled") do |capsule|
      capsule.concurrency = 1
      capsule.queues = ["mb2_throttled", 1]
    end
  end
```

We are not sure why we are seeing this behaviour. It seems to have gotten worse since the 7.3.0 release. We're seeing this behaviour multiple times a day now. Please let us know if you need any additional information.

**Comment by mperham** [comment]
No clue.

**Comment by RemcodM** [comment]
This is still happening to us in production. Today, we were able to actually see this happening from the Sidekiq interface:

![Scherm_afbeelding 2024-08-19 om 14 37 53](https://github.com/user-attachments/assets/dc1f365c-a085-45bc-95ea-d5245c05a7b3)

2 times the same job with the same job ID appearing and running across two different threads.

We are still not sure what causes this behavior to happen. Somehow the job pick-up does not seem to be transactional?

**Comment by Sybe** [comment]
Today we found another example with 3 times the same job ID running at the same time. These are jobs taking a bit longer. We see about 2 minutes between the first and third start.

**Comment by mperham** [comment]
This is how retries work. They can restart the same job over several minutes. How do you know this isn't normal behavior?

**Comment by Sybe** [comment]
These are not retries. We see extra starts of the same job ID before the first job fails, so multiple instances of the same job run at the same time. 
For retries we expect the job to first fail, and then restart with the same job ID, but having no moment where the same job ID runs on multiple threads similtaneously.

**Comment by mperham** [comment]
You'll need to provide more data to debug. Please provide the list of any sidekiq-* gems/extensions you're using and your full initializer. If you can repro it with a simple app, that would be best.

**Comment by RemcodM** [comment]
@mperham We had some kind of revelation this morning as what is happening.

We run Sidekiq in Kubernetes. That means, the pods are volatile: They appear and disappear relatively often. To ensure that all jobs run correctly, we use a termination time-out. That means: We ensure all our jobs run within 2 minutes, and when Kubernetes decides it is time to terminate a Sidekiq process, we give it 2 minutes to finish the jobs it is running. So far so good.

The problem now appears to be that once we send the termination signal, the jobs that are still running on the Sidekiq proces (and still have 2 minutes left to finish) are already started on different Sidekiq process, causing them to run simultaneously. This behavior seems to have changed relatively recently. Before, the jobs would only be started in a new process if somehow they did not make the 2 minutes deadline (which they should) or if they failed within the 2 minute deadline. Now, the termination signal itself is enough to restart alle the jobs running inside that Sidekiq proces.

I am not sure if this is something you recognize from a recent change. But I imagine this in unwanted behaviour; we specify `-t 115` for a reason: We want Sidekiq to have roughly 2 minutes to finish up running jobs. That should not cause Sidekiq to immediatly restart all (still running) jobs on a different pod.

**Comment by mperham** [comment]
I would suggest you lower that timeout from 120 seconds back down to the default of 30 seconds and see if that fixes the issue.

**Comment by mperham** [comment]
How are you starting Sidekiq in k8s? Show me the actual exec line.

**Comment by RemcodM** [comment]
The actual command line:
```
command: ["bundle", "exec", "sidekiq", "-C", "config/sidekiq.yml", "-t", "115"]
```

Obviously `sidekiq.yml` as shown above.

**Comment by RemcodM** [comment]
> I would suggest you lower that timeout from 120 seconds back down to the default of 30 seconds and see if that fixes the issue.

We could obviously try is bringing it back to the default of 30 seconds would help, however, that has consequences for us since not every job is finished within that time frame. We have had it at 115 seconds for years and only encountered problems recently. Also, our scaling behavior has not changed within Kubernetes... So it is not the case that we are suddenly experiencing much more dropouts.

It is kinda interesting that we see the job start in a new Sidekiq process **immediately** after the termination signal is sent. It remains running on the terminating instance, but also gets started on a new instance. This does not happen after 10, 20 of even 30 seconds in the termination state, it happens **immediately**. I am almost certain that didn't happen in previous versions, although I am not entirely certain at which version this problem started.

**Comment by mperham** [comment]
Ok, that command is good. Some people start Sidekiq via bash, which can trap signals, breaking the shutdown process and orphaning jobs.

I would try to determine when the behavior broke. Go back to 7.2.4 and see if the good behavior resumes.

**Comment by Sybe** [comment]
We went back to both 7.2.4 and after that to 7.1.6, we still see the wrong behaviour.

**Comment by mperham** [comment]
That's definitely not expected behavior but bugs can be located anywhere or be caused by several subsystems interacting badly. The fact that a Sidekiq version downgrade is not solving the problem tells me that the root cause is likely not Sidekiq itself but another component that was originally upgraded, e.g. the redis-client gem, a Sidekiq 3rd party plugin, etc. We need more data to narrow the search.

**Comment by mscrivo** [comment]
> We need more data to narrow the search.

agreed, working on it. One additional piece of info I just discovered at least in our case. This seems to be affecting long running jobs that are running while the sidekiq deployment is restarted. I cleared out the queue of the 25+ items and left one in the retry queue. Upon starting up, it correctly picked up the 1 item and started working on it. After an hour or so, it was restarted and when it came back up, boom there were two copies running at the same time now. And and and additional entry in the retry page with this error:

`ThreadError: can't alloc thread` error

**Comment by mperham** [comment]
`ThreadError: can't alloc thread` is a very bad sign and indicates you've got a resource issue somewhere but I have no idea how to debug that issue. What are you doing that's strange or unusual? What's your concurrency?

**Comment by mk-dhia** [comment]
Hello,

we have the same issue with job getting started multiple time. it seems to be related to jobs that get killed (and may be when pod start the shutdown process, and get killed before it ends the process)

We have a monthly job that get killed frequently (because it last for 30 minutes (node get killed by aws) , and use a lot of RAM  sometime it get oom killed). the job is started in batch and do upload of files and as post processing in the batch we close the file. what we notice is when a job run a second time after the upload has been closed (as all job finshed from batch point of view) the second run fail. this is why we noticed it. may be there are other occurrences but if the job is idempotent we will not notice it.

we start to see this problem on september 1 st and on almost each run we did since. we use  sidekiq 7.3.1/7.2.0(pro)/7.2.2(ent)
redis_client 0.22.2 and we were using on august  7.2.0/7.2.0/7.2.2 redis_client 0.18

here more logs for a test run i did on september 25:

![image](https://github.com/user-attachments/assets/9433a7fa-9509-420b-9f65-11749aa7367c)

![image](https://github.com/user-attachments/assets/4e1c4f27-313d-40a9-9de2-33a644c178cb)

![image](https://github.com/user-attachments/assets/0e0aa8f8-0147-42c2-a87c-08e1b7fead8d)

delay in the execution is may be due to the latency on the queue that wher ethe job is processed. it's possible that all 3 instance as been put back in queue at the same time but due to delay we see launch 29 minutes apart
![image](https://github.com/user-attachments/assets/ff746d70-48a7-4027-adac-bb9a2f4d09d8)


as first guess i would think of a race condition between the gracefull shutdown that renqueue the jobs if they do not finish within the timeout and superfetch of cleanup_the_dead

**Comment by mscrivo** [comment]
We hit this issue again, having multiple same JID's running at the same time that severely strained our DB because of deadlocks running the same queries. It appears the `ThreadError: can't alloc thread` is the root cause for the duplicated jos and I believe it's because of our use of multi-threaded dogtstats we use. What's unique about dogstats and sidekiq is that in addition to our main dogstats thread pool, we have sidekiq pro's metrics hooked up to dogstatsd as well, so that creates even more threads: 

```rb
Sidekiq::Pro.dogstatsd = -> do
  Datadog::Statsd.new(ENV['DOGSTATSD_IP'], 8125, namespace: 'sidekiq')
end
```

We're going to try and set dogstatsd to be single-threaded and see if the problem goes away.

**Comment by fatkodima** [comment]
@mperham 

We see a similar behavior. 

2 days ago a single job was enqueued and yesterday we noticed there were **1420** duplicates of the same job running at the same time (same jid, same created_at for the job)! The job failed on some `nil` error in its body and was at 16-th retry when we noticed that.

Any ideas?

We are on the almost latest sidekiq. 
Initializer
```ruby
# frozen_string_literal: true

Rails.application.reloader.to_prepare do
  require 'sidekiq/middleware/current_attributes'
  Sidekiq::CurrentAttributes.persist(Current)

  Sidekiq.configure_server do |config|
    config.redis = { url: ENV.fetch('SIDEKIQ_REDIS_URL') }

    config.super_fetch!
    config.client_middleware do |chain|
      chain.add SidekiqExt::TagsMiddleware
    end

    config.server_middleware do |chain|
      chain.add SidekiqExt::ShardingMiddleware::Server

      chain.add SidekiqExt::Metrics
      chain.add SidekiqExt::MonitoringMiddleware
      chain.add SidekiqExt::ExecutionContextMiddleware
    end

    config[:job_logger] = SidekiqExt::JobLogger

    config.on(:fork) do
      ActiveRecordConnectionPoolMonitor.new('Sidekiq').start
      LaunchDarklyClientManager.initialize_client
    end

    config.on(:shutdown) do
      LaunchDarklyClientManager.shutdown_client
    end

    if ENV['SIDEKIQ_DOSESPOT_CONCURRENCY'] && config[:queues].delete('dosespot')
      # don't process dosespot jobs in the default capsule
      config.capsules['default'].queues.delete('dosespot')
      config.capsules['default'].weights.delete('dosespot')

      config.capsule("dosespot") do |cap|
        cap.concurrency = ENV['SIDEKIQ_DOSESPOT_CONCURRENCY'].to_i
        cap.queues = ['dosespot']
      end
    end

    unless ENV['DISABLE_SIDEKIQ_PAIRED_RANDOM_CONCURRENCY']
      if config[:queues].any?(/,/)
        raise "A paired random sidekiq capsule can only be used when queues are defined without weights."
      end

      config.capsule("paired_random") do |cap|
        cap.concurrency = config[:concurrency]
        cap.queues = config[:queues].without('low').map{ "#{_1},1" }
      end
    end
  end

  Sidekiq::Job::ClassMethods.include(SidekiqExt::UniqueContext)

  Sidekiq.configure_client do |config|
    config.redis = { url: ENV.fetch('SIDEKIQ_REDIS_URL') }

    config.client_middleware do |chain|
      # Adds tags (like current shard) to jobs.
      chain.add SidekiqExt::TagsMiddleware
    end
  end
end

# https://gist.github.com/marcotc/39b0d5e8100f0f4cd4d38eff9f09dcd5
Sidekiq.default_job_options = { retry: 19 } # 6.5 days

unless Rails.env.test?
  Sidekiq::Client.reliable_push!
  Sidekiq::Enterprise.unique!
end
```

Job's body roughly:
```ruby
class GenerateChartingNotesJob
  include Sidekiq::Job
  sidekiq_options queue: 'default'

  def perform(args)
    do_something
  end
end
```

**Comment by mperham** [comment]
@fatkodima That's concerning but I don't have any idea what might be wrong. I'd suggest you open your own issue, link to this one, and provide any more debugging / logging for that specific JID in your logs and any "nearby" errors that might be associated.

PS I don't think it's a good idea to put everything inside `to_prepare`. Rails will run it on every request and that itself may cause problems.

https://api.rubyonrails.org/classes/ActiveSupport/Reloader.html

I think you just need to stringify this Class and `to_prepare` can be removed completely:

```ruby
  require 'sidekiq/middleware/current_attributes'
  Sidekiq::CurrentAttributes.persist("Current")
```

**Comment by fatkodima** [comment]
`to_prepare` runs only once in production, and only in development it is running on each request. This is still needed, because there are some other app classes are used, besides `Current`.

I think I figured out the possible problem, will open a separate issue.

**Comment by stephanmoneybird** [comment]
**Reproducible locally — root cause identified in `Capsule#stop` + SuperFetch interaction**

I've been able to consistently reproduce this locally with a single Redis instance (no replicas), confirming this is not a Redis replication lag issue.

**Environment:** Sidekiq 8.0.10, sidekiq-pro 8.0.3, sidekiq-ent 8.0.3, multiple capsules, `super_fetch!` enabled.

**Reproduction:**

A test job that logs its execution to Redis:

```ruby
class SuperfetchTestJob
  include Sidekiq::Job
  sidekiq_options queue: 'superfetch_test', retry: 0

  def perform(test_id)
    Sidekiq.redis do |conn|
      conn.rpush("superfetch_test:#{test_id}:executions", {
        pid: Process.pid, jid: jid, started_at: Time.now.iso8601
      }.to_json)
    end
    sleep 30
  end
end
```

Steps:
1. Start two Sidekiq processes: `bundle exec sidekiq -q superfetch_test -c 1 -t 115`
2. Enqueue the job: `SuperfetchTestJob.perform_async("test1")`
3. Wait a few seconds for the job to start, then send SIGTERM to one process
4. The other process picks up and starts executing the same JID within seconds

**Note:** This reproduction requires multiple capsules per process. In our case, the initializer creates a "throughput" capsule alongside the default capsule, both listening on the same queue. The bug triggers because the capsule without active workers shuts down first. A single-capsule setup will not reproduce this issue.

```
Total executions of JID: 2
Unique PIDs: 2 (91589, 91590)

  Execution 1: PID=91589 started=2026-04-16T11:56:46+02:00
  Execution 2: PID=91590 started=2026-04-16T11:56:52+02:00
```

**Root cause:**

The issue is the interaction between [`Capsule#stop`](https://github.com/sidekiq/sidekiq/blob/main/lib/sidekiq/capsule.rb#L53-L55) and SuperFetch's `bulk_requeue` when multiple capsules are configured.

`Capsule#stop` unconditionally calls `fetcher.bulk_requeue([])`:

```ruby
def stop
  fetcher&.bulk_requeue([])
end
```

This is called from [`Manager#stop`'s `ensure` block](https://github.com/sidekiq/sidekiq/blob/main/lib/sidekiq/manager.rb#L65-L66) — meaning it runs even when a capsule has no active workers and returns early (line 58). SuperFetch's `bulk_requeue` ignores the empty argument and moves **all** jobs from the private queue back to the public queue.

The critical detail: SuperFetch private queues are keyed by **process identity**, not capsule. Multiple capsules in the same process share the same private queue for a given queue name. When a capsule with no workers shuts down first, its `bulk_requeue` requeues jobs that another capsule is still actively processing.

**Timeline:**
1. SIGTERM → `Manager#stop` runs per capsule in separate threads
2. Capsule with no active workers → `return if @workers.empty?` → `ensure` → `capsule.stop` → `bulk_requeue([])`
3. SuperFetch moves ALL jobs from the shared private queue back to the public queue
4. Other capsule's processor is still running the job
5. Another Sidekiq process picks it up → **duplicate execution**

This explains why the duplicate starts within seconds of SIGTERM (not after the `-t` timeout).

**Fix:**

Skip `bulk_requeue` when called with an empty in-progress list (from `Capsule#stop`). `Manager#hard_shutdown` passes the real in-progress jobs list and continues to work as before. Orphaned jobs in private queues are recovered by `cleanup_the_dead` after heartbeat expiry.

```ruby
if defined?(Sidekiq::Pro::SuperFetch)
  module SuperFetchBulkRequeueFix
    def bulk_requeue(inprogress = [])
      return if Array(inprogress).empty?

      super
    end
  end

  Sidekiq::Pro::SuperFetch.prepend(SuperFetchBulkRequeueFix)
end
```

**Comment by mperham** [comment]
I'm wondering why `Capsule#stop` calls `bulk_requeue` at all? `Manager#hard_shutdown` does it for us after the `-t` timeout. Calling with `[]` is a no-op for basic_fetch. We shouldn't be running `bulk_requeue` until all Processor threads are dead.

**Comment by mperham** [comment]
I see the issue, one Manager per Capsule so line 106 is called once per capsule. We can remove the call in Capsule#stop and change super_fetch return if empty but I am worried about leaking Redis data. Is it possible a race condition could lead to a job not being pushed back from a private queue, e.g. with a perfectly timed Sidekiq::Shutdown exception? I suppose super_fetch would rescue that job and clean up the data so maybe this is not a dealbreaker.

**Comment by stephanmoneybird** [comment]
The `Sidekiq::Shutdown` timing shouldn't be an issue here. In `hard_shutdown`, `bulk_requeue(jobs)` runs **before** `processor.kill` raises `Sidekiq::Shutdown`, so jobs are already moved back to the public queue by the time the exception is raised.

For the edge case where a job is in the private queue but the process dies before `hard_shutdown` can run (e.g. OOM kill, `SIGKILL`), `cleanup_the_dead` already handles this — it recovers jobs from dead processes' private queues after heartbeat expiry. That's the existing safety net and it doesn't depend on `Capsule#stop`'s `bulk_requeue` call.

**Comment by mperham** [comment]
I believe this is an alternative fix ^^^^^. Can you run `main` and verify?

**Comment by stephanmoneybird** [comment]
Confirmed. I applied the same change locally and the reproduction script no longer triggers duplicate execution.

**Comment by mperham** [comment]
8.1.3 is now available.
